Create rule S6790: Disallow using string references (`react/no-string-refs`) (#3187)
SonarSource/rspec: rules/S6790/javascript/rule.adoc

== Why is this an issue?

React Refs provide a way to access DOM nodes or React elements created in the render method.

> Older React versions allowed the ref attribute to be a string, like "textInput", later accessible as `this.refs.textInput`. This is considered legacy due to multiple reasons:

Reviewer: ```suggestion
Older React versions allowed the ref attribute to be a string, like `"textInput"`, later accessible as `this.refs.textInput`. This is considered legacy code due to multiple reasons:
```

* String `refs` make React slower as they force React to keep track of what component is currently executing.

> * String `refs` are not composable: If a library puts a ref on the passed child, the user can't put another ref on it.

Reviewer: ```suggestion
* String `refs` are not composable: if a library puts a ref on the passed child, the user can't put another ref on it.
```

> * The owner for a string `ref` is determined by the currently executing component.

Reviewer: ```suggestion
* The owner of a string `ref` is determined by the currently executing component.
```

[source,javascript,diff-id=1,diff-type=noncompliant]
----

> var Hello = createReactClass({

Reviewer: ```suggestion
const Hello = createReactClass({
```

  componentDidMount: function() {

>     var component = this.refs.hello; // Noncompliant

Reviewer: ```suggestion
    const component = this.refs.hello; // Noncompliant
    // ...
```

  },
  render: function() {
    return <div ref="hello">Hello, world.</div>;
  }
});
----

> Instead, reference callbacks should be used. These do not have the limitations mentioned above. When the DOM node is added to the screen, React will call the `ref` callback with the DOM node as the argument. When that DOM node is removed, React will call your `ref` callback with `null`. One should return `undefined` from the `ref` callback.

Reviewer: ```suggestion
Instead, reference callbacks should be used. These do not have the limitations mentioned above. When the DOM node is added to the Web page, React will call the `ref` callback with the DOM node as the argument. When that DOM node is removed, React will call your `ref` callback with `null`. One should return `undefined` from the `ref` callback.
```

[source,javascript,diff-id=1,diff-type=compliant]
----

> var Hello = createReactClass({

Reviewer: ```suggestion
const Hello = createReactClass({
```

  componentDidMount: function() {

>     var component = this.hello;

Reviewer: ```suggestion
    const component = this.hello;
    // ...
```
